File: LeetCode/2021/Unique Paths.java
class Solution 
{
    private void printMatrix(int[][] matrix)
    {
        for (int i = 0; i < matrix.length; i++)
        {
            for (int j = 0; j < matrix[i].length; j++)
            {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.print("\n");
        }
    }
    
    private void initMatrix(int[][] num_moves)
    {
        for (int i = 0; i < num_moves.length; i++)
        {
            num_moves[i][num_moves[0].length-1] = 1;
        }
        for (int i = 0; i < num_moves[0].length; i++)
        {
            num_moves[num_moves.length-1][i] = 1;
        }
    }
    
    private void fillMatrix(int[][] num_moves)
    {
        for (int i = num_moves.length-2; i >= 0; i--)
        {
            for (int j = num_moves[0].length-2; j >= 0; j--)
            {
                num_moves[i][j] = num_moves[i+1][j] + num_moves[i][j+1];
            }
        }
    }
    
    public int uniquePaths(int m, int n)
    {
        // idea: set up an mxn matrix to refer to for previous values
        int[][] num_moves = new int[m][n];
        initMatrix(num_moves);
        fillMatrix(num_moves);
        return num_moves[0][0];
    }
}

File: LeetCode/2021/Unique Paths.cs
public class Solution 
{
    private void PrintMatrix(int[,] matrix)
    {
        for (int i = 0; i < matrix.GetLength(0); i++)
        {
            for (int j = 0; j < matrix.GetLength(1); j++)
            {
                System.Console.Write(matrix[i, j] + ", ");
            }
            System.Console.WriteLine("");
        }
    }
    
    public int UniquePaths(int m, int n) 
    {
        int [,] uniquePaths = new int[m, n];
        for (int i = 0; i < m; i++)
        {
            uniquePaths[i, 0] = 1;
        }
        for (int i = 0; i < n; i++)
        {
            uniquePaths[0, i] = 1;
        }
        for (int i = 1; i < m; i++)
        {
            for (int j = 1; j < n; j++)
            {
                uniquePaths[i, j] = uniquePaths[i-1, j] + uniquePaths[i, j-1];
            }
        }
        // PrintMatrix(uniquePaths);
        return uniquePaths[m-1, n-1];
    }
}
